src/replication.rs

```diff
@@ -40,22 +45,73 @@ pub struct ReplicationState {
     snapshot_cache: (DateTime<Utc>, Cache),
     db_pool: Arc<PgPool>,
     storage: Arc<FileSystem>,
-    // clients: Arc<HashMap<String, ObjectServiceClient<tonic::transport::Channel>>>,
+    clients: MutexF<HashMap<String, CachedClient<ObjectServiceClient<tonic::transport::Channel>>>>
 }
 
+#[derive(Clone, Debug)]
+struct CachedClient<T> {
+    id: String,
+    client: MutexF<T>
+}
+
+impl <T> CachedClient<T> {
+    pub fn new(client: T) -> Self {
+        Self {
+            id: Uuid::new_v4().to_string(),
+            client: MutexF::new(client)
+        }
+    }
+
+    pub fn id(&self) -> &String {
+         &self.id
+    }
+
+    pub fn lock(&self) -> impl Future<Output = MutexGuard<T>> {
+        self.client.lock()
+    }
+}
+
+
 impl ReplicationState {
     pub fn new(cache: Arc<Mutex<Cache>>, config: Arc<Config>, db_pool: Arc<PgPool>, storage: Arc<FileSystem>) -> Self {
         let snapshot_cache = cache.lock().unwrap().clone();
         let snapshot_cache = (Utc::now(), snapshot_cache);
 
-        Self { cache, config, snapshot_cache, db_pool, storage }
+        Self { cache, config, snapshot_cache, db_pool, storage, clients: MutexF::new(HashMap::new()) }
+    }
+
+    async fn cached_client(&self, url: &String) -> Result<CachedClient<ObjectServiceClient<tonic::transport::Channel>>> {
+        let cc = {
+            let mut guard = self.clients.lock().await;
+            let connect_url = url.clone();
+            match guard.entry(url.clone()) {
+                Entry::Occupied(entry) => (*entry.get()).clone(),
+                Entry::Vacant(entry) => {
+                    let client = ObjectServiceClient::connect(connect_url)
+                        .await
+                        .map_err(ReplicationError::TonicTransportError)?;
+                    entry.insert(CachedClient::new(client)).clone()
+                }
+            }
+        };
+        Ok(cc)
     }
 }
 
 async fn replicate_public_key(inner: &ReplicationState, public_key: &String, url: String) -> Result<usize> {
     let batch = datastore::replication_object_uuids(&inner.db_pool, public_key, inner.config.replication_batch_size).await?;
     let batch_size = batch.len();
-    let mut client = ObjectServiceClient::connect(url).await?;
+    let backoff_strategy = ExponentialBackoff::default();
+
+    // Retry the attempt to connect up to 10 times (the default with the `backoff` crate) with exponential backoff.
+    let cached_client = retry(backoff_strategy, || async {
```

Comment:
This is pretty much what I was originally thinking as well. That's why I had the `clients` field commented out. I didn't implement it right away because I was trying to think through a better scheme. The biggest problem here is that `replicate_public_key` gets called for all remote public keys concurrently and we await the result of joining all those futures. This means that if a single remote_public_key has an object store that isn't running, every batch would wait for that single remote_public_key to fail the full backoff. It would result in reduced throughput for all the other remote_public_keys. The other thing that is unfortunate is having the `clients` field in the struct requires the use of a `Mutex`. We know that a specific client connection is only ever used by a single future in each `replicate_iteration` call. In addition to the above, having a lockless scheme would be ideal.

Would something like this work?

Inside of `replicate` holds the client map and it gets passed into `replicate_iteration`. The correct remote connection gets passed into `replicate_public_key`. It would be `Option<tonic::transport::Channel>`. If it's present it is used, if it is `None` a new connection is created. Something would be returned in order to signify if it should be removed or added back to the `clients` hashmap. This would mean that the `None` could be used inside of `replicate_iteration` in order to just skip some `remote_public_keys`. This skipping would amount to backing off. Some sort of time might also need to be held in the map so we know how long we should skip it for.

This is just one strategy I thought up. There could be others that accomplish the same task, but the aim would be these two requirements:
- lockless
- backing off of a subset of connections doesn't impact the replication of other keys, other than for the loops where we are trying those bad connections again every so often.

The backoff can also get lengthy pretty fast too. I image there could be cases where we're trying to replicate to a key that's just not fully setup yet, or there's firewall problems. In which case we could probably do something like 30s, 1m, 2m, 4m, 8m, 16m, 32m, 32m, 32m, 32m, ...

Reply:
@scirner22 That's a lot of useful context. I think the approach you outlined should work. To @piercetrey-figure point, do you think it's necessary to implement backoff for the `put` call as well?

Reply:
I think utilizing the same back off functionality for both of those would be great. Specifically in the `put` failure case - we should probably log it at the error level if it's a non connection related problem. Any non connection related error would be a bug.